lib/fs.js

```diff
@@ -2479,3 +2479,454 @@ Object.defineProperty(fs, 'SyncWriteStream', {
   set: internalUtil.deprecate((val) => { SyncWriteStream = val; },
                               'fs.SyncWriteStream is deprecated.', 'DEP0061')
 });
+
+
+// The Promises API
+
+async function* readEverything(fd) {
+  let loops = 0;
+  let eof = false;
+  let position = 0;
+  const buffer = Buffer.allocUnsafe(kReadFileBufferLength);
+  while (!eof) {
+    if (loops > 5) process.exit(0);
+    const { length } =
+      await fs.promises.read(fd, buffer, 0, buffer.length, position);
+    if (length < buffer.length) {
+      eof = true;
+      return buffer.slice(0, length);  // Ended
+    } else {
+      position += length;
+      yield buffer.slice(0, length);
+    }
+  }
+}
```

Comment:
@mcollina ... take note of this function. It is an async generator used in an async iterator within `fs.promises.readFile()`. Significantly less complicated than the current callback based implementation.